fendermint/app/options/src/upgrade.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::path::PathBuf;
+
+use clap::{Args, Subcommand};
+
+#[derive(Args, Debug)]
+pub struct UpgradeArgs {
+    /// Path to the upgrade_info JSON file.
+    #[arg(long)]
+    pub upgrade_file: PathBuf,
+
+    #[command(subcommand)]
+    pub command: UpgradeCommands,
+}
+
+#[derive(Subcommand, Debug)]
+pub enum UpgradeCommands {
+    AddUpgrade(AddUpgrade),
+}
+
+#[derive(Args, Debug)]
+pub struct AddUpgrade {
+    /// the height at which to schedule the upgrade
+    #[arg(long)]
+    pub height: u64,
+
+    /// the application version the upgrade will update to
+    #[arg(long)]
+    pub new_app_version: u64,
+
+    /// the required cometbft version for the upgrade
+    #[arg(long)]
+    pub cometbft_version: String,
+
+    /// whether this is a required upgrade or not. A required upgrade
+    /// will cause the node to freeze and not process any more blocks
+    /// if it does not have the corresponding Upgrade defined to
+    /// migrate to that version
+    #[arg(long)]
+    pub required: bool,
```

Comment:
Just thinking when an upgrade wouldn't be required. I know we talked about upgrades where you can download the new software and replace the existing one with it _now_, because the binary is backwards compatible, and it will apply the upgrade when the time comes. Why the upgrade list would know this though..

OTOH if some nodes update the `app_version` then everyone must do it, because it's part of the ledger. So an optional upgrade must not update the app version.

Reply:
I'm also not sure how an optional upgrade would be handled if the node with the upgrade is started later than the height. Nothing will trigger that upgrade, so we don't know if it's been applied or not?